hpedockerplugin/cmd/cmd_createshare.py

```diff
@@ -0,0 +1,273 @@
+                                self._share_args['vfsIPs'] = [(ips[0],
+                                                               subnet)]
+                                return self._create_share()
+            except Exception:
+                pass
+            raise ex
+
+    # If default FPG is full, it raises exception
+    # EtcdMaxSharesPerFpgLimitException
+    def _get_default_available_fpg(self):
+        fpg_name = self._get_current_default_fpg_name()
+        fpg_info = self._fp_etcd.get_fpg_metadata(self._backend,
+                                                  self._share_args['cpg'],
+                                                  fpg_name)
+        if fpg_info['share_cnt'] >= share.MAX_SHARES_PER_FPG:
+            raise exception.EtcdMaxSharesPerFpgLimitException(
+                fpg_name=fpg_name)
+        return fpg_info
+
+    def _get_current_default_fpg_name(self):
+        cpg_name = self._share_args['cpg']
+        try:
+            backend_metadata = self._fp_etcd.get_backend_metadata(
+                self._backend)
+            default_fpgs = backend_metadata.get('default_fpgs')
+            if default_fpgs:
+                default_fpg = default_fpgs.get(cpg_name)
+                if default_fpg:
+                    return default_fpg
+            raise exception.EtcdDefaultFpgNotPresent(cpg=cpg_name)
+        except exception.EtcdMetadataNotFound:
+            raise exception.EtcdDefaultFpgNotPresent(cpg=cpg_name)
+
+
+class CreateShareOnExistingFpgCmd(CreateShareCmd):
+    def __init__(self, file_mgr, share_args):
+        super(CreateShareOnExistingFpgCmd, self).__init__(file_mgr,
+                                                          share_args)
+
+    def execute(self):
+        fpg_name = self._share_args['fpg']
+        with self._fp_etcd.get_fpg_lock(self._backend, fpg_name):
+            try:
+                # Specified FPG may or may not exist. In case it
+                # doesn't, EtcdFpgMetadataNotFound exception is raised
+                fpg_info = self._fp_etcd.get_fpg_metadata(
+                    self._backend, self._share_args['cpg'], fpg_name)
+                self._share_args['vfs'] = fpg_info['vfs']
+                # Only one IP per FPG is supported at the moment
+                # Given that, list can be dropped
+                subnet_ips_map = fpg_info['ips']
+                subnet, ips = next(iter(subnet_ips_map.items()))
+                self._share_args['vfsIPs'] = [(ips[0], subnet)]
+                self._create_share()
+            except exception.EtcdMetadataNotFound as ex:
+                # Assume it's a legacy FPG, try to get details
+                fpg_info = self._get_legacy_fpg()
+
+                # CPG passed can be different than actual CPG
+                # used for creating legacy FPG. Override default
+                # or supplied CPG
+                self._share_args['cpg'] = fpg_info['cpg']
```

Comment:
Why do we need to over right the cpg name? I agree that inspect on share will show correct CPG as meta data is corrected but original request was for a particular CPG. In case of k8s SC will have passed cpg information which can not be modified and contrast to we are overriding this with correct value. 

Instead I suggest to raise an exception telling user passed cpg name doesn't match with associated existing cpg.